examples/thesis_work/collision_probability.py

```diff
@@ -0,0 +1,261 @@
+    
+    return (1.0 / (2.0 * sigma_i * delta_i)) * (term1 - term2 - term3 + term4)
+
+def build_1d_collision_probability_matrix(regions: List[Tuple[float, float]]) -> np.ndarray:
+    """
+    Build the collision probability matrix for 1D geometry.
+    
+    Parameters
+    ----------
+    regions : List[Tuple[float, float]]
+        List of (sigma, width) tuples for each region
+        
+    Returns
+    -------
+    np.ndarray
+        Collision probability matrix P
+    """
+    n_regions = len(regions)
+    P = np.zeros((n_regions, n_regions))
+    
+    # self-collision probabilities
+    for i in range(n_regions):
+        sigma_i, delta_i = regions[i]
+        P[i, i] = calculate_1d_self_collision_probability(sigma_i, delta_i)
+    
+    # collision probabilities between different regions
+    for i in range(n_regions):
+        sigma_i, delta_i = regions[i]
+        
+        # position of region boundaries
+        x = np.zeros(n_regions + 1)
+        x[0] = 0.0
+        for k in range(n_regions):
+            x[k+1] = x[k] + regions[k][1]
+        
+        for j in range(n_regions):
+            if i != j:
+                sigma_j, delta_j = regions[j]
+                
+                # optical path between regions
+                if j > i:
+                    tau_ij = sum(regions[k][0] * regions[k][1] for k in range(i+1, j))
+                else:  # j < i
+                    tau_ij = sum(regions[k][0] * regions[k][1] for k in range(j+1, i))
+                
+                P[i, j] = calculate_1d_collision_probability(sigma_i, sigma_j, delta_i, delta_j, tau_ij)
+    
+    return P
+
+def calculate_3d_collision_probability_matrix(mesh, num_rays: int = 100):
+    """
+    Calculate collision probabilities for a 3D mesh using ray tracing with the new API.
+    
+    Parameters
+    ----------
+    mesh : openmc.RegularMesh
+        Mesh dividing the geometry into voxels
+    num_rays : int, optional
+        Number of rays to use for Monte Carlo estimation
+        
+    Returns
+    -------
+    tuple
+        (P, sigma_total, sigma_scatter) where:
+        P : np.ndarray - Collision probability matrix
+        sigma_total : np.ndarray - Total cross section for each voxel
+        sigma_scatter : np.ndarray - Scattering cross section for each voxel
+    """
+    lower_left = np.array(mesh.lower_left)
+    upper_right = np.array(mesh.upper_right)
+    dimensions = np.array(mesh.dimension)
+
+    n_voxels = np.prod(dimensions)
+
+    # initialize matrices
+    P = np.zeros((n_voxels, n_voxels))
+    sigma_total = np.zeros(n_voxels)
+    sigma_scatter = np.zeros(n_voxels)
+
+    def voxel_bounds(index):
+        i = index % dimensions[0]
+        j = (index // dimensions[0]) % dimensions[1]
+        k = index // (dimensions[0] * dimensions[1])
+        min_bound = lower_left + (upper_right - lower_left) * np.array([i, j, k]) / dimensions
+        max_bound = min_bound + (upper_right - lower_left) / dimensions
```

Comment:
[performance] with the above definition of `delta`
```suggestion
        min_bound = lower_left + delta * np.array([i, j, k]) 
        max_bound = min_bound + delta
```